File: backend/app/schemas/matching.py
from uuid import UUID

from pydantic import BaseModel


class CommonTag(BaseModel):
    tag: str
    weight_mine: str
    weight_theirs: str


class MatchUser(BaseModel):
    id: UUID
    username: str
    avatar_url: str | None = None
    bio: str | None = None


class MatchResult(BaseModel):
    user: MatchUser
    score: float
    is_viewed: bool
    common_tags: list[CommonTag]


class MatchList(BaseModel):
    results: list[MatchResult]
    total: int
    page: int
    limit: int
    pages: int


class SimilarityResponse(BaseModel):
    user_id: UUID
    score: float
    common_tags: list[CommonTag]


class ViewOut(BaseModel):
    viewer_id: UUID
    viewed_id: UUID
    is_viewed: bool


class ContactOut(BaseModel):
    from_user_id: UUID
    to_user_id: UUID
    status: str

File: frontend/src/api/matching.ts
import { api } from './client'

export interface CommonTag {
  tag: string
  weight_mine: string
  weight_theirs: string
}

export interface MatchUser {
  id: string
  username: string
  avatar_url: string | null
  bio: string | null
}

export interface MatchResult {
  user: MatchUser
  score: number
  is_viewed: boolean
  common_tags: CommonTag[]
}

export interface MatchList {
  results: MatchResult[]
  total: number
  page: number
  limit: number
  pages: number
}

export interface Tag {
  id: string
  name: string
  category: string | null
}

export interface TagList {
  items: Tag[]
  total: number
  page: number
  limit: number
  pages: number
}

export interface SimilarityResponse {
  user_id: string
  score: number
  common_tags: CommonTag[]
}

export const matchingApi = {
  list: (page = 1, limit = 20) =>
    api.get<MatchList>('/matching', { params: { page, limit } }).then(r => r.data),

  getSimilarity: (userId: string) =>
    api.get<SimilarityResponse>(`/matching/${userId}`).then(r => r.data),
}

export const tagsApi = {
  search: (q: string, page = 1, limit = 10) =>
    api.get<TagList>('/tags', { params: { q, page, limit } }).then(r => r.data),

  create: (name: string) =>
    api.post<Tag>('/tags', { name }).then(r => r.data),
}
